File: maximum-units-on-a-truck.ts
{
    const maximumUnits = function(boxTypes, truckSize) {
        let c = 0;

        boxTypes = boxTypes.sort((a,b) => b[1] - a[1])

        for (let i = 0;i < boxTypes.length; i++) {
            if (truckSize <= 0) {
                return c
            }

            const [size, weight] = boxTypes[i]

            if (truckSize > size) {
                c += size * weight
                truckSize -= size
            } else {
                c += truckSize * weight;
                truckSize -= size
            }
        }

        return c;
    };


    console.log(maximumUnits([[1,3],[2,2],[3,1]], 4))
}

File: maximum-units-on-a-truck.rs
pub fn maximum_units(mut box_types: Vec<Vec<i32>>, truck_size: i32) -> i32 {
    let mut c = 0;
    let mut size = truck_size;

    box_types.sort_by(|a,b| b[1].cmp(&a[1]));

    for i in 0..box_types.len() {
        if size <= 0 {
            return c;
        } else {
            if size > box_types[i][0] {
                c += box_types[i][0] * box_types[i][1];
                size -= box_types[i][0];
            } else {
                c += size * box_types[i][1];
                return  c;
            }
        }
    }

    return c;
}